lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java

```diff
@@ -147,69 +203,299 @@ public void markDocumentEventsProcessedOrFailed(
+                            entity, update.originalProcessingDate));
+
+                    savedEvents.add(event);
+                }
+            }
+        }
+
+        LightblueBulkDataResponse bulkResponse;
+
+        try {
+            bulkResponse = lightblue.bulkData(insertAndUpdateEvents);
+        } catch (LightblueBulkResponseException e) {
+            // If some failed, that's okay. We have to iterate through responses either way.
+            // We'll check for errors then.
+            bulkResponse = e.getBulkResponse();
+        }
+
+        Iterator<LightblueDocumentEvent> eventsIterator = savedEvents.iterator();
+        Iterator<LightblueDataResponse> responsesIterator = bulkResponse.getResponses().iterator();
+
+        while (eventsIterator.hasNext()) {
+            if (!responsesIterator.hasNext()) {
+                throw new IllegalStateException("Mismatched number of requests and responses! " +
+                        "Events looked like: <{}>. Responses looked like");
+            }
+
+            LightblueDataResponse response = responsesIterator.next();
+            LightblueDocumentEvent event = eventsIterator.next();
+            DocumentEventEntity entity = event.wrappedDocumentEventEntity();
+
+            if (response instanceof LightblueErrorResponse) {
+                LightblueErrorResponse errorResponse = (LightblueErrorResponse) response;
+
+                // Likely transient failure; leave event alone to be tried again later.
+                if (errorResponse.hasDataErrors() || errorResponse.hasLightblueErrors()) {
+                    if (logger.isWarnEnabled()) {
+                        List<Error> errors = new ArrayList<>();
+
+                        Error[] lightblueErrors = errorResponse.getLightblueErrors();
+                        DataError[] dataErrors = errorResponse.getDataErrors();
+
+                        if (lightblueErrors != null) {
+                            Collections.addAll(errors, lightblueErrors);
+                        }
+
+                        if (dataErrors != null) {
+                            Collections.addAll(errors, Arrays.stream(dataErrors)
+                                    .flatMap(dataError -> dataError.getErrors().stream())
+                                    .toArray(Error[]::new));
+                        }
+
+                        List<String> errorStrings = errors.stream()
+                                .map(e -> "Code: " + e.getErrorCode() + ", " +
+                                        "Context: " + e.getContext() + ", " +
+                                        "Message: " + e.getMsg())
+                                .collect(Collectors.toList());
+
+                        logger.warn("Event update failed. Will not process. Event was: <{}>. " +
+                                "Errors: <{}>", event, errorStrings);
+                    }
+
+                    eventsIterator.remove();
```

Comment:
We remove all locks after we return the document events, since they are marked processing with timestamp (this happens in the try-with-resources block)